src/v/cluster/decommissioning_monitor.cc

```diff
@@ -0,0 +1,222 @@
+    if (cfg.type() != raft::configuration_type::simple) {
+        co_return;
+    }
+
+    // check if node was actually removed from raft0 configuration
+    // removing node from the decommissioned list finishes the decommissioning
+    // process
+    std::erase_if(_decommissioned, [&cfg](model::node_id id) {
+        return !cfg.contains_broker(id);
+    });
+
+    co_return;
+}
+
+bool is_replica_set_up_to_date(
+  const cluster::topic_configuration_assignment& cfg,
+  const model::ntp& ntp,
+  const std::vector<model::broker_shard>& requested_replica_set) {
+    auto it = std::find_if(
+      cfg.assignments.begin(),
+      cfg.assignments.end(),
+      [pid = ntp.tp.partition](const partition_assignment& pas) {
+          return pid == pas.id;
+      });
+    vassert(
+      it != cfg.assignments.end(),
+      "Reallocated partition {} have to exists ",
+      ntp);
+
+    return it->replicas == requested_replica_set;
+}
+
+ss::future<>
+decommissioning_monitor::reallocate_replica_set(reallocation_meta& meta) {
+    // ask partition allocator for reallocation of
+    // replicas
+    vlog(clusterlog.debug, "reallocating partition {}", meta.ntp);
+
+    if (!meta.new_assignment) {
+        meta.new_assignment = co_await _allocator.invoke_on(
+          partition_allocator::shard,
+          [current = meta.current_assignment](
+            partition_allocator& allocator) mutable {
+              return allocator.reallocate_decommissioned_replicas(current);
+          });
+    }
+    // was unable to assign partitition replicas, do nothing
+    if (!meta.new_assignment) {
+        co_return;
+    }
+
+    vlog(
+      clusterlog.debug,
+      "moving replicas of {} to {}",
+      meta.ntp,
+      meta.new_assignment->get_assignments().begin()->replicas);
+    std::error_code ec
+      = co_await _topics_frontend.local().move_partition_replicas(
+        meta.ntp,
+        meta.new_assignment->get_assignments().begin()->replicas,
+        model::timeout_clock::now() + _retry_timeout);
+
+    vlog(clusterlog.debug, "moving replicas result: {}", ec.message());
+    auto it = _topics.local().topics_map().find(
+      model::topic_namespace_view(meta.ntp));
+    // topic was deleted, we are done with assignment
+    if (it == _topics.local().topics_map().end()) {
+        meta.is_finished = true;
+        co_return;
+    }
+    meta.is_finished = is_replica_set_up_to_date(
+      it->second,
+      meta.ntp,
+      meta.new_assignment->get_assignments().begin()->replicas);
+}
+
+bool is_in_replica_set(
+  const std::vector<model::broker_shard>& replicas,
+  const absl::flat_hash_set<model::node_id>& ids) {
+    auto it = std::find_if(
+      std::cbegin(replicas),
+      std::cend(replicas),
+      [&ids](const model::broker_shard& bs) {
+          return ids.contains(bs.node_id);
+      });
+
+    return it != std::cend(replicas);
+}
+
+absl::flat_hash_set<model::node_id> ids_set(std::vector<model::node_id> ids) {
```

Comment:
could generalize this

Reply:
actually it isn't used, I will remove this.